File: java/src/main/java/org/sparky/model/builders/BlockBuilder.java
package org.sparky.model.builders;

import org.sparky.model.*;
import org.sparky.model.keys.ConstantKey;
import org.sparky.model.keys.IndexerKey;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Created by lloyd on 17/10/2015.
 */
public class BlockBuilder {

    private String name;
    private VariableBuilder nameBuilder;
    private boolean isIndexer = false;

    private List<BlockBuilder> blocks = new ArrayList<>();
    private Map<String, Map<Boolean, EvaluatableBuilder>> rules = new HashMap<>();
    private TableBuilder table = null;

    public BlockBuilder withBlock(BlockBuilder block){
        blocks.add(block);
        return this;
    }

    public BlockBuilder withName(String name){
        this.name = name;
        return this;
    }

    public BlockBuilder withName(VariableBuilder variableBuilder){
        this.nameBuilder = variableBuilder;
        return this;
    }

    public BlockBuilder withIndexer(){
        isIndexer = true;
        return this;
    }

    public BlockBuilder withTable(TableBuilder table){
        this.table = table;
        return this;
    }

    public BlockBuilder withRule(boolean overrideFlag, String name, EvaluatableBuilder rule){
        if(!rules.containsKey(name)){
            rules.put(name, new HashMap<>());
        }

        rules.get(name).put(overrideFlag, rule);
        return this;
    }

    public Block build(Table parentTable) {
        Table t = table == null ? parentTable : table.build();

        List<Block> b = blocks.stream().map(bl -> bl.build(t)).collect(Collectors.toList());

        Map<Key, Rule> kvps = new HashMap<>();
        for(String key : rules.keySet()){

            Key k = new ConstantKey(key);
            Rule r = null;

            if(rules.get(key).size() > 1){
                r = rules.get(key).get(true).build();
            } else {
                r = rules.get(key).values().iterator().next().build();
            }
            kvps.put(k, r);
        }

        if(isIndexer){
            return new IndexerBlock(new IndexerKey(t), b, kvps, t);
        }

        return new Block(new ConstantKey(name), b, kvps, t);
    }
}

File: net/Sparky/Model/Builder/BlockBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sparky.Model.Builder
{
    public class BlockBuilder
    {
        private string _name;
        private IEvaluatableBuilder _nameBuilder;
        private bool _isIndexer = false;

        private List<BlockBuilder> _children = new List<BlockBuilder>();
        private IDictionary<string, IDictionary<bool, IEvaluatableBuilder>> _rules = new Dictionary<string, IDictionary<bool, IEvaluatableBuilder>>();
        private TableBuilder _table;

        public BlockBuilder WithBlock(BlockBuilder block)
        {
            _children.Add(block);
            return this;
        }

        public BlockBuilder WithName(string name)
        {
            _name = name;
            return this;
        }

        public BlockBuilder WithName(IEvaluatableBuilder nameBuilder)
        {
            _nameBuilder = nameBuilder;
            return this;
        }

        public BlockBuilder WithIndexer()
        {
            _isIndexer = true;
            return this;
        }

        public BlockBuilder WithTable(TableBuilder tb)
        {
            _table = tb;
            return this;
        }

        public BlockBuilder WithRule(bool overrideFlag, string name, IEvaluatableBuilder rule)
        {
            if (!_rules.ContainsKey(name))
            {
                _rules.Add(name, new Dictionary<bool, IEvaluatableBuilder>());
            }
            _rules[name].Add(overrideFlag, rule);
            return this;
        }

        public Block Build(Table parent)
        {
            Table t = _table == null ? parent : _table.Build();

            List<Block> b = _children.Select(c => c.Build(t)).ToList();

            IDictionary<IKey, Rule> kvps = new Dictionary<IKey, Rule>();
            foreach(var k in _rules.Keys)
            {
                IKey ky = new ConstantKey(k);
                Rule r = null;
                if(_rules[k].Count > 1)
                {
                    r = _rules[k][true].Build();
                }
                else
                {
                    r = _rules[k].Values.First().Build();
                }
                kvps.Add(ky, r);
            }

            if (_isIndexer)
            {
                return new IndexerBlock(new IndexerKey(t), b, kvps, t);
            }
            return new Block(new ConstantKey(_name), b, kvps, t);
        }
    }
}
